auth/authzserver/resource_server.go

```diff
@@ -0,0 +1,115 @@
+package authzserver
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"mime"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/flyteorg/flytestdlib/config"
+
+	"github.com/coreos/go-oidc"
+	"github.com/flyteorg/flyteadmin/auth"
+
+	authConfig "github.com/flyteorg/flyteadmin/auth/config"
+	"github.com/flyteorg/flyteadmin/auth/interfaces"
+	"github.com/flyteorg/flyteidl/gen/pb-go/flyteidl/service"
+	"golang.org/x/oauth2"
+)
+
+type ResourceServer struct {
+	signatureVerifier oidc.KeySet
+}
+
+func (r ResourceServer) ValidateAccessToken(ctx context.Context, expectedAudience, tokenStr string) (interfaces.IdentityContext, error) {
+	raw, err := r.signatureVerifier.VerifySignature(ctx, tokenStr)
+	if err != nil {
+		return nil, err
+	}
+
+	claimsRaw := map[string]interface{}{}
+	if err = json.Unmarshal(raw, &claimsRaw); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal user info claim into UserInfo type. Error: %w", err)
+	}
+
+	return verifyClaims(expectedAudience, claimsRaw)
+}
+
+func doRequest(ctx context.Context, req *http.Request) (*http.Response, error) {
+	client := http.DefaultClient
+	if c, ok := ctx.Value(oauth2.HTTPClient).(*http.Client); ok {
+		client = c
+	}
+	return client.Do(req.WithContext(ctx))
+}
+
+func unmarshalResp(r *http.Response, body []byte, v interface{}) error {
+	err := json.Unmarshal(body, &v)
+	if err == nil {
+		return nil
+	}
+	ct := r.Header.Get("Content-Type")
+	mediaType, _, parseErr := mime.ParseMediaType(ct)
+	if parseErr == nil && mediaType == "application/json" {
+		return fmt.Errorf("got Content-Type = application/json, but could not unmarshal as JSON: %v", err)
+	}
+	return fmt.Errorf("expected Content-Type = application/json, got %q: %v", ct, err)
+}
+
+func getJwksForIssuer(ctx context.Context, issuerBaseURL url.URL) (oidc.KeySet, error) {
+	u, err := url.Parse(auth.OAuth2MetadataEndpoint)
```

Comment:
In case of using google as the idp, the link https://accounts.google.com/.well-known/oauth-authorization-server does not exist so the `getJwksForIssuer` is failing when configured the `baseUrl` as `https://accounts.google.com`. Can we fall back to use `OIdCMetadataEndpoint` when this is failing? So this link does exist `https://accounts.google.com/.well-known/openid-configuration`

Reply:
Google can't be used as the external OAuth2 Authorization Server... their server only protects other Google Endpoints. You can setup/use [GCP Identity Platform](https://cloud.google.com/identity-platform) instead (I do not have instructions for how to do that)...
If you follow the guide [here](https://github.com/flyteorg/flyte/pull/773), it takes you through setting up Google Idp just for OpenID Connect while continuing to use FlyteAdmin's own certs to act as the OAuth2 Authorization Server

Reply:
We actually need to use Google as authentication server because of gcloud service accounts. What want to do for authentication is just to verify who is actually `who` based on ID token. Regarding authorisation, when we know what the user it is we can fetch custom roles from google (it is possible to add custom role) and implement a very thin layer of access control. In this way we don't need have a separate service to store access policies.

Reply:
That's exactly what the guide takes you through... setting up OpenID Connect with Google IdP... have you tried that?
Let's chat over slack (@haytham) and I would be happy to help you debug and get it up and running

Reply:
Thanks. I managed to get it running but it required me to make some code changes (hack). I will create a PR for these changes.